Fine tune array-array intersection algorithm #63

ppiotrow opened the pull request:
The aim of the commit is to verify optimal value for Util.unsignedIntersect2by2 parameter.
Currently it's is param=64
The commit contains first version of benchmarking the `galloping` and `local` algorithm variants.
For param={"10","15","20","25","30","35", "40", "50","60", "90"} it generates five different examples:
 * small container S
 * big contained B (B.length = param * S.size)
The container values are generated by Uniform Distribution in version1.
The examples are deterministic, random seed is hardcoded.

lemire commented:
@ppiotrow 

Great. Do you have some numbers to share?

Can you check the spelling of ``UtilBechmark.java``?

ppiotrow commented:
## number of examples `local` algorithm was faster than `galloping` for given param.
<img width="531" alt="wins" src="https://cloud.githubusercontent.com/assets/3899229/20502132/096cdc46-b03d-11e6-8034-ce5ff29a90cb.png">

## for every experiment comparison how better was `local` than `galloping` [ops/s] 
<img width="1003" alt="comparison" src="https://cloud.githubusercontent.com/assets/3899229/20502138/0ed1ace8-b03d-11e6-9ac9-c059c93bdf10.png">

ppiotrow commented:
Let's treat this as Work In Progress. I'm going to generate container content with different distributions (currently uniform). I uploaded this to gather your suggestions as you probably know best what could be worth to check.

lemire commented:
@ppiotrow 

So your results suggest that setting the parameter to 32 instead of 64 could be beneficial.

ppiotrow commented:
At least for Uniform distribution this is true. Of course it is worth to increase granularity and check differences for `param = {30,31,32,33,34,35}`.

lemire commented:
A non-uniform distribution that we sometimes use is this one: 

https://github.com/lemire/JavaFastPFOR/blob/master/src/main/java/me/lemire/integercompression/synth/ClusteredDataGenerator.java

ppiotrow commented:
Thanks for that generator, I'll try it. 
Also for documentation purposes this is copy of python notebook I used to obtain images from benchmark results https://gist.github.com/ppiotrow/d5f2a2f813a7ca8202480be579ef021b

lemire commented:
I like your Python script. It is neat and simple.

ppiotrow commented:
As the benchmarked code is used only in `and` operations between two array containers it makes no sense to benchmark containers larger than 1024. Am I right? Currently in the benchmark there are defined `SMALLEST_ARRAY = 50` and `BIGGEST_ARRAY = 2^16`. I think those values should be `2` and `1024`.

lemire commented:
Array containers can contain between 1 and 4096 integers (inclusively).

ppiotrow commented:
Now the benchmark uses both uniform and clustered distributions in following way:
small(uniform)    with big(uniform)
small(clustered) with big(uniform)
small(uniform)    with big(clustered)
small(clustered) with big(clustered)
I also fixed the data so now generated arrays are between 2 and 4096 elements.
Bellow there are updated results of the benchmark. The gist was also updated.

For every experiment comparison how better was local than galloping [ops/s]
<img width="984" alt="zrzut ekranu 2016-12-03 o 12 49 05" src="https://cloud.githubusercontent.com/assets/3899229/20859165/3a1c013c-b957-11e6-9588-45941aed1fb6.png">
Number of examples local algorithm was faster than galloping for given param.
<img width="521" alt="zrzut ekranu 2016-12-03 o 12 49 17" src="https://cloud.githubusercontent.com/assets/3899229/20859167/3a427600-b957-11e6-962b-fb9b19276ab6.png">
Number of examples local algorithm was faster than galloping for given param (only **uniform** distributions)
<img width="515" alt="zrzut ekranu 2016-12-03 o 12 49 29" src="https://cloud.githubusercontent.com/assets/3899229/20859168/3a43495e-b957-11e6-9eb3-b8cf0b6b3d38.png">
Number of examples local algorithm was faster than galloping for given param (only **clustered** distributions)
<img width="507" alt="zrzut ekranu 2016-12-03 o 12 49 39" src="https://cloud.githubusercontent.com/assets/3899229/20859166/3a4124a8-b957-11e6-8af8-d3ce441bbc6f.png">

ppiotrow commented:
The next steps I suggest are
* analyze the cases for `param=30` where local was few times faster than galloping and vice versa,
* perform benchmark for `param = {25,26.....33,34,35}` as it looks like there is best param value.

ppiotrow commented:
![unknown](https://cloud.githubusercontent.com/assets/3899229/20869336/c7c1ceb2-ba70-11e6-907b-616b28645123.png)
<img width="514" alt="all" src="https://cloud.githubusercontent.com/assets/3899229/20869309/ef0f2d08-ba6f-11e6-8fa7-80c02c971945.png">
<img width="512" alt="uniform" src="https://cloud.githubusercontent.com/assets/3899229/20869308/ef0ed86c-ba6f-11e6-9b92-6b3effa0716f.png">
<img width="520" alt="clustered" src="https://cloud.githubusercontent.com/assets/3899229/20869307/ef0d3214-ba6f-11e6-8697-c98bebb9b904.png">
Based on the first picture which shows detailed comparison of throughput of `local` and `galloping` approach it looks like from `param=26` we tend to have better results for galloping. Even if `local` is sometimes better (for param={26,27...} it is rather win ratio around 1-15%. It is meaningless given that `galloping` wins 1-70% and the big differences are quite frequent.

lemire commented:
@ppiotrow Speaking for myself, if you were to change the default based on these results, I think we could merge it and issue a new release. Given that the current default was ad hoc, chances are good that it would be generally beneficial.

ppiotrow commented:
Last result with more experiments
![unknown](https://cloud.githubusercontent.com/assets/3899229/20945338/2119f2ae-bc07-11e6-9e91-8328c2c17eee.png)
I will set threshold to 25.

lemire commented:
@ppiotrow  You may want to make your status in Roaring public https://github.com/orgs/RoaringBitmap/people/ppiotrow

lemire commented:
LGTM

lemire commented:
Public release RoaringBitmap-0.6.28
